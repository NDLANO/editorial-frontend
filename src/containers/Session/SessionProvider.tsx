/**
 * Copyright (c) 2021-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { Auth0DecodedHash } from "auth0-js";
import { createContext, Dispatch, ReactNode, SetStateAction, useContext, useState } from "react";
import { useNavigate } from "react-router-dom";
import {
  clearAccessTokenFromLocalStorage,
  getAccessToken,
  getAccessTokenPersonal,
  personalAuthLogout,
  setAccessTokenInLocalStorage,
} from "../../util/authHelpers";
import handleError from "../../util/handleError";
import { decodeToken, isValid } from "../../util/jwtHelper";
import { toLogin } from "../../util/routeHelpers";

const SessionContext = createContext<[SessionState, Dispatch<SetStateAction<SessionState>>] | undefined>(undefined);

interface Props {
  children: ReactNode;
  initialValue?: SessionState;
}

interface UserData {
  name?: string;
  permissions?: string[];
  ndlaId?: string;
}

interface SessionState {
  user: UserData;
  authenticated: boolean;
  userNotRegistered: boolean;
}

export const initialState: SessionState = {
  user: {},
  authenticated: false,
  userNotRegistered: true,
};

export interface SessionProps {
  userName?: string;
  ndlaId: string | undefined;
  userPermissions?: string[];
  authenticated: boolean;
  userNotRegistered: boolean;
  login: (authResult: Auth0DecodedHash) => void;
  logout: (federated: boolean, returnToLogin?: boolean) => void;
}

export const getSessionStateFromLocalStorage = (): SessionState => {
  const token = getAccessToken();
  const isAccessTokenPersonal = getAccessTokenPersonal();
  if (isValid(token) && isAccessTokenPersonal) {
    const decodedToken = decodeToken(token);
    return {
      user: {
        name: decodedToken?.["https://ndla.no/user_name"],
        ndlaId: decodedToken?.["https://ndla.no/ndla_id"],
        permissions: decodedToken?.permissions,
      },
      authenticated: true,
      userNotRegistered: false,
    };
  }
  return initialState; // Return initial state if token is undefined
};

export const SessionProvider = ({ children, initialValue = initialState }: Props) => {
  const sessionState = useState<SessionState>(initialValue);
  return <SessionContext value={sessionState}>{children}</SessionContext>;
};

export const useSession = (): SessionProps => {
  const sessionContext = useContext(SessionContext);
  if (sessionContext === undefined) {
    throw new Error("useSession must be used within a SessionProvider");
  }
  const [session, setSession] = sessionContext;
  const navigate = useNavigate();

  const setAuthenticated = (authenticated: boolean) => setSession((s) => ({ ...s, authenticated }));
  const setUserNotRegistered = (userNotRegistered: boolean) => setSession((s) => ({ ...s, userNotRegistered }));
  const setUserData = (user: UserData) => setSession((s) => ({ ...s, user }));

  const login = (authResult: Auth0DecodedHash) => {
    try {
      const decoded = isValid(authResult.accessToken ?? null) ? decodeToken(authResult.accessToken!) : undefined;
      const permissions = decoded?.permissions ?? [];
      const scope = decoded?.scope ?? "";
      const combinedPermissions = [...permissions, ...scope.split(" ")];
      const uniquePermissions = [...new Set(combinedPermissions)];

      if (decoded && uniquePermissions.some((permission) => permission.includes(":"))) {
        if (authResult.state) window.location.href = authResult.state;
        setAuthenticated(true);
        setUserData({
          name: decoded["https://ndla.no/user_name"],
          permissions: uniquePermissions,
        });
        setAccessTokenInLocalStorage(authResult.accessToken!, true);
        navigate("/", { replace: true });
      } else {
        setUserNotRegistered(true);
        navigate(`${toLogin()}/failure`, { replace: true });
      }
    } catch (e) {
      handleError(e);
    }
  };

  const logout = (federated: boolean, returnToLogin = false) => {
    try {
      setAuthenticated(false);
      setUserData({});
      personalAuthLogout(federated, returnToLogin);
      clearAccessTokenFromLocalStorage();
    } catch (e) {
      handleError(e);
    }
  };

  return {
    userName: session.user.name,
    ndlaId: session.user.ndlaId,
    userPermissions: session.user.permissions,
    authenticated: !!session.authenticated,
    userNotRegistered: !!session.userNotRegistered,
    login,
    logout,
  };
};
