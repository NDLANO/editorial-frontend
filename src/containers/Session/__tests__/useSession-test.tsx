/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { ElementType, ReactNode } from "react";
import { MemoryRouter } from "react-router";
import { act, renderHook } from "@testing-library/react";
import { getSessionStateFromLocalStorage, SessionProvider, useSession } from "../SessionProvider";

const navigateMock = vi.fn();
vi.mock("react-router", async (original) => {
  const actual: object = await original(); // get actual module
  return {
    ...actual,
    useNavigate: () => navigateMock,
  };
});

const authResult = {
  accessToken:
    "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiIsImtpZCI6Ik9FSTFNVVU0T0RrNU56TTVNekkyTXpaRE9EazFOMFl3UXpkRE1EUXlPRFZDUXpRM1FUSTBNQSJ9.eyJodHRwczovL25kbGEubm8vbmRsYV9pZCI6Im1ObXJ1N1lvU0lqUGVwT3JrTjRNTWhsVSIsImh0dHBzOi8vbmRsYS5uby91c2VyX25hbWUiOiJUZXN0IHVzZXIiLCJodHRwczovL25kbGEubm8vY2xpZW50X2lkIjoiSks0Uk8zVTciLCJpc3MiOiJodHRwczovL25kbGEuZXUuYXV0aDAuY29tLyIsInN1YiI6Imdvb2dsZS1vYXV0aDJ8MTIzNDIzNDIzMTMyMTM1NjEyMyIsImF1ZCI6Im5kbGFfc3lzdGVtIiwiaWF0IjoxNTA5OTcwMDI3LCJleHAiOjE1MDk5NzcyMjcsImF6cCI6IldVMEtyNENEa3JNMHVMOXhZZUZWNGNsOUdhMXZCM0pZIiwic2NvcGUiOiJzb21lOnBlcm1pc3Npb24iLCJwZXJtaXNzaW9ucyI6WyJzb21lOnBlcm1pc3Npb24iXX0.n7YdEm0FypAz8NdEDLcVyDFXHnNG_Zo7T7RNt7WRAa4",
};

afterEach(() => {
  localStorage.clear();
});

interface Props {
  initialValue?: any;
}

const createWrapper = (Wrapper: ElementType, props: Props) => {
  return ({ children }: { children?: ReactNode }) => (
    <MemoryRouter>
      <Wrapper {...props}>{children}</Wrapper>
    </MemoryRouter>
  );
};

describe("getSessionStateFromLocalStorage", () => {
  it("returns a default initial state if token is not available in localStorage", () => {
    localStorage.removeItem("access_token");
    expect(getSessionStateFromLocalStorage()).toMatchSnapshot();
  });

  it("returns a default initial state if accessToken is not personal", () => {
    localStorage.setItem("access_token", authResult.accessToken);
    localStorage.removeItem("access_token_personal");
    expect(getSessionStateFromLocalStorage()).toMatchSnapshot();
  });

  it("returns a decoded access token if everything is ok", () => {
    localStorage.setItem("access_token", authResult.accessToken);
    localStorage.setItem("access_token_personal", "true");
    const nextState = getSessionStateFromLocalStorage();

    expect(nextState).toMatchSnapshot();
  });
});

describe("useSession", () => {
  afterEach(() => {
    navigateMock.mockClear();
  });

  afterAll(() => {
    vi.clearAllMocks();
  });
  test("correctly sets access token and access_token_personal on successful login", async () => {
    const { result } = renderHook(() => useSession(), {
      wrapper: createWrapper(SessionProvider, {}),
    });
    act(() => {
      result.current.login(authResult);
    });
    expect(navigateMock).toHaveBeenCalledTimes(1);
    expect(navigateMock).toHaveBeenCalledWith("/", { replace: true });
    expect(localStorage.getItem("access_token")).toEqual(authResult.accessToken);
    expect(localStorage.getItem("access_token_personal")).toEqual("true");
  });

  test("returns initial state when called without initial values", () => {
    const { result } = renderHook(() => useSession(), {
      wrapper: createWrapper(SessionProvider, {}),
    });
    expect(result.current).toMatchSnapshot();
  });

  test("returns initial state when called with initial values", () => {
    localStorage.setItem("access_token", authResult.accessToken);
    localStorage.setItem("access_token_personal", "true");
    const { result } = renderHook(() => useSession(), {
      wrapper: createWrapper(SessionProvider, {
        initialValue: getSessionStateFromLocalStorage(),
      }),
    });

    expect(result.current).toMatchSnapshot();
  });

  test("nulls all values when logout is called", () => {
    const { result } = renderHook(() => useSession(), {
      wrapper: createWrapper(SessionProvider, {}),
    });
    act(() => {
      result.current.login(authResult);
    });
    expect(result.current).toMatchSnapshot();
    act(() => {
      result.current.logout(true);
    });
    expect(localStorage.getItem("access_token")).toBe(null);
    expect(localStorage.getItem("access_token_expires_at")).toBe(null);
    expect(localStorage.getItem("access_token_personal")).toBe(null);
    expect(result.current).toMatchSnapshot();
  });
});
