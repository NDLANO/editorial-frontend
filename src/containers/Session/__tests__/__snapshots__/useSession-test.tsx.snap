// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSessionStateFromLocalStorage returns a decoded access token if everything is ok 1`] = `
Object {
  "authenticated": true,
  "user": Object {
    "name": "Test user",
    "ndlaId": "mNmru7YoSIjPepOrkN4MMhlU",
    "permissions": Array [
      "some:permission",
    ],
  },
  "userNotRegistered": false,
}
`;

exports[`getSessionStateFromLocalStorage returns a default initial state if accessToken is not personal 1`] = `
Object {
  "authenticated": false,
  "user": Object {},
  "userNotRegistered": true,
}
`;

exports[`getSessionStateFromLocalStorage returns a default initial state if token is not available in localStorage 1`] = `
Object {
  "authenticated": false,
  "user": Object {},
  "userNotRegistered": true,
}
`;

exports[`useSession nulls all values when logout is called 1`] = `
Object {
  "authenticated": true,
  "login": [Function],
  "logout": [Function],
  "ndlaId": undefined,
  "userName": "Test user",
  "userNotRegistered": true,
  "userPermissions": Array [
    "some:permission",
  ],
}
`;

exports[`useSession nulls all values when logout is called 2`] = `
Object {
  "authenticated": false,
  "login": [Function],
  "logout": [Function],
  "ndlaId": undefined,
  "userName": undefined,
  "userNotRegistered": true,
  "userPermissions": undefined,
}
`;

exports[`useSession returns initial state when called with initial values 1`] = `
Object {
  "authenticated": true,
  "login": [Function],
  "logout": [Function],
  "ndlaId": "mNmru7YoSIjPepOrkN4MMhlU",
  "userName": "Test user",
  "userNotRegistered": false,
  "userPermissions": Array [
    "some:permission",
  ],
}
`;

exports[`useSession returns initial state when called without initial values 1`] = `
Object {
  "authenticated": false,
  "login": [Function],
  "logout": [Function],
  "ndlaId": undefined,
  "userName": undefined,
  "userNotRegistered": true,
  "userPermissions": undefined,
}
`;
